### PR TITLE
Use multiprecision::msb() in GasMeter and remove FixedHash dependency

### PR DESCRIPTION
**Pull request by axic**
Part of #7794.

### CODE REVIEW

File: libevmasm/GasMeter.cpp [hunk 1]
```diff
@@ -182,7 +180,11 @@ GasMeter::GasConsumption GasMeter::estimateMax(AssemblyItem const& _item, bool _
 		case Instruction::EXP:
 			gas = GasCosts::expGas;
 			if (u256 const* value = classes.knownConstant(m_state->relativeStackElement(-1)))
-				gas += GasCosts::expByteGas(m_evmVersion) * (32 - (h256(*value).firstBitSet() / 8));
+			{
+				// Note: msb() counts from 0.
+				auto const significant_byte_count = (boost::multiprecision::msb(*value) + 1 + 7) / 8;
```
Comment (ekpyron) [inline libevmasm/GasMeter.cpp]: What will ``boost::multiprecision::msb`` for ``u256(0)``? IIRC it will throw, right? We don't want that, do we?

Comment (axic) [reply libevmasm/GasMeter.cpp]: Will it throw? I just couldn't find any documentation on it.

Comment (ekpyron) [reply libevmasm/GasMeter.cpp]: boost's documentation is really horrible :-)...

Comment (ekpyron) [reply libevmasm/GasMeter.cpp]: https://www.boost.org/doc/libs/1_72_0/libs/multiprecision/doc/html/boost_multiprecision/tut/gen_int.html
says something about " Throws a std::domain_error if x <= 0. "

Comment (ekpyron) [reply libevmasm/GasMeter.cpp]: But I haven't checked...

Comment (axic) [reply libevmasm/GasMeter.cpp]: If it does, we don't have a clear test for it.

Update: added a test case to trigger and indeed it throws.
```
Dynamic exception type: boost::wrapexcept<std::range_error>
std::exception::what: No bits were set in the operand.
```

File: libevmasm/GasMeter.cpp [hunk 2]
```diff
@@ -182,7 +180,16 @@ GasMeter::GasConsumption GasMeter::estimateMax(AssemblyItem const& _item, bool _
 		case Instruction::EXP:
 			gas = GasCosts::expGas;
 			if (u256 const* value = classes.knownConstant(m_state->relativeStackElement(-1)))
```
Comment (ekpyron) [inline libevmasm/GasMeter.cpp:182]: Wouldn't it be simpler to have
```
if (u256 const* value = classes.knownConstant(m_state->relativeStackElement(-1)))
  if (*value)
  {
    // Note: msb() counts from 0
    unsigned significant_byte_count =  (boost::multiprecision::msb(*value) + 1 + 7) / 8;
    gas += GasCosts::expByteGas(m_evmVersion) * significant_byte_count;
  }
```

Or maybe even with ``if (init; cond)`` :-):

```
if (
  u256 const* value = classes.knownConstant(m_state->relativeStackElement(-1)));
  value && *value
)
{
  // Note: msb() counts from 0
  unsigned significant_byte_count =  (boost::multiprecision::msb(*value) + 1 + 7) / 8;
  gas += GasCosts::expByteGas(m_evmVersion) * significant_byte_count;
}
```

Comment (axic) [reply libevmasm/GasMeter.cpp:182]: Feel free to take the PR over :)

My goal was to shrink FixedHash. More of that shrinking is here https://github.com/ethereum/solidity/tree/fixedhash

**Comment by axic** [comment]
Rebased after #8159.

**Review by ekpyron — DISMISSED** [review]

File: libevmasm/GasMeter.cpp [hunk 3]
```diff
@@ -182,7 +180,14 @@ GasMeter::GasConsumption GasMeter::estimateMax(AssemblyItem const& _item, bool _
...
-				gas += GasCosts::expByteGas(m_evmVersion) * (32 - (h256(*value).firstBitSet() / 8));
+			{
+				if (*value)
+				{
+					// Note: msb() counts from 0 and throws on 0 as input.
+					unsigned const significant_byte_count  = (boost::multiprecision::msb(*value) + 1 + 7) / 8;
```
Comment (leonardoalt) [inline libevmasm/GasMeter.cpp]: Camel style

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**